File: Recursion_problems/MinCostForTickets.java

// Problem link - https://leetcode.com/problems/minimum-cost-for-tickets/description/

package Recursion_problems;

import java.util.*;

public class MinCostForTickets {
    public static int[] daysArr;
    public static int[] costsArr;
    public static Set<Integer> s;

    private static int findMinimumCost(int day){
        // base case1: if 'd'th day is greater than last day of days array
        if(day > daysArr[daysArr.length - 1]) return 0;

        // Recursive case: if dth day doesn't exist in set 
        if(!s.contains(day)) return findMinimumCost(day+1);

        // recursively calculate minimim cost: f(day) = min(costs[0]+f(d+1), costs[1]+f(d+8), costs[2]+f(d+30))
        int ans = Math.min(costsArr[0]+findMinimumCost(day+1), Math.min(costsArr[1]+findMinimumCost(day+7), costsArr[2]+findMinimumCost(day+30)));

        return ans;
    }
    public static int mincostTickets(int[] days, int[] costs) {
        daysArr = days;
        costsArr = costs;
        s = new HashSet<Integer>();

        for(int d : days){
            s.add(d);
        }
        return findMinimumCost(days[0]);
    }

    public static void main(String[] args) {
        int[] days = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 30, 31};
        int[] costs = {2, 7, 15};

        // int[] days = {1,4,6,7,8,20};
        // int[] costs = {2, 7, 15};
        int ans = mincostTickets(days, costs);
        System.out.println(ans);
    }
}

File: 04.Recursion_problems/MinCostForTickets.js

// Problem link - https://leetcode.com/problems/minimum-cost-for-tickets/description/


// Approach - using Recursion

// global variables 'dArr' for days & 'cArr' for costs array
let dArr;
let cArr;

// using to set store days
let s;

function findMinimumCostRec(d){
    // base case1: if 'd'th day is greater than last day of days array
    if(d > dArr[dArr.length - 1]) return 0;
    // Recursive case: if dth day doesn't exist in set 
    if(!s.has(d)) return findMinimumCostRec(d + 1);

    // recursive case: f(d) = min(costs[0]+f(d+1), costs[1]+f(d+8), costs[2]+f(d+30))
    let ans = Math.min(cArr[0]+findMinimumCostRec(d+1), cArr[1]+findMinimumCostRec(d+7), cArr[2]+findMinimumCostRec(d+30));

    return ans;
}

let mincostTickets = function(days, costs) {
    s = new Set(days);
    dArr = days;
    cArr = costs;
    return findMinimumCostRec(days[0]);
};

// let days = [1,4,6,7,8,20];
// let costs = [2,7,15];

let days = [1,2,3,4,5,6,7,8,9,10,30,31];
let costs = [2,7,15];
let ans = mincostTickets(days, costs);
console.log(ans);